src/camera.nim

```diff
@@ -98,26 +104,11 @@ proc newPerspectiveCamera*(a, d: float32; transf = Transformation.id): Camera {.
     Camera(kind: ckPerspective, transf: transf, aspect_ratio: a, distance: d)
 
 proc fire_ray*(cam: Camera; pixel: Point2D): Ray {.inline.} = 
+    var ray: Ray
     case cam.kind
     of ckOrthogonal:
-        apply(cam.transf, newRay(newPoint3D(-1, (1 - 2 * pixel.u) * cam.aspect_ratio, 2 * pixel.v - 1), eX))
+        ray = newRay(newPoint3D(-1, (1 - 2 * pixel.u) * cam.aspect_ratio, 2 * pixel.v - 1), eX)
     of ckPerspective:
-        apply(cam.transf, newRay(newPoint3D(-cam.distance, 0, 0), newVec3(cam.distance, (1 - 2 * pixel.u) * cam.aspect_ratio, 2 * pixel.v - 1)))
-
-    
-type ImageTracer* = object
-    image*: HdrImage
-    camera*: Camera
-
-proc fire_ray*(im_tr: ImageTracer; x, y: int, pixel = newPoint2D(0.5, 0.5)): Ray {.inline.} =
-    im_tr.camera.fire_ray(newPoint2D((x.float32 + pixel.u) / im_tr.image.width.float32, 1 - (y.float32 + pixel.v) / im_tr.image.height.float32))
-
-proc fire_all_rays*(im_tr: var ImageTracer) = 
-    for x in 0..<im_tr.image.height:
-        for y in 0..<im_tr.image.width:
-            discard im_tr.fire_ray(x, y)
-            let 
-                r = (1 - exp(-float32(x + y)))
-                g = y / im_tr.image.height
-                b = pow((1 - x / im_tr.image.width), 2.5)
-            im_tr.image.setPixel(x, y, newColor(r, g, b))
+        ray = newRay(newPoint3D(-cam.distance, 0, 0), newVec3(cam.distance, (1 - 2 * pixel.u) * cam.aspect_ratio, 2 * pixel.v - 1))
+
+    ray.transform(cam.transf)
```

Comment:
ok i like it, it's the same for every kind of camera so it's not necessary to have it in every case